app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,159 @@
+import mozilla.components.support.utils.SafeIntent
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.APP_ICON
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.ERROR
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.CUSTOM_TAB
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.LINK
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Source.UNKNOWN
+
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppAllStartup.Type.HOT
+
+/**
+ * Tracks application startup source, type, and whether or not activity has savedInstance to restore
+ * the activity from. Sample metric = [source = COLD, type = APP_ICON, hasSavedInstance = false]
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppAllStartup] and finally report them on Activity's onResume() function.
+ */
+@Suppress("TooManyFunctions")
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    private var onCreateData: AppAllStartup? = null
+    private var onRestartData: Pair<Type, Boolean>? = null
+    private var onNewIntentData: Source? = null
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, false),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, true),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    /**
+     * we are not setting [Source] in this method since source is derived from a [SafeIntent].
+     * therefore source either gets set in oncreate() or onNewIntent().
+     */
+    fun onHomeActivityOnRestart() {
+        onRestartData = Pair(HOT, false)
+    }
+
+    /**
+     * we are only setting [Source] in this method since source is derived from a [SafeIntent].
+     * other metric fields are set in onCreate() and onRestart()
```

Comment:
This is also false and we should move it to the function body.

tbh, I'm also fine not having this comment because:
- the metrics tracking code makes that assumption somewhat clear
- it adds complexity to this function: now the dev starts thinking about a broader context than, "lifecycle function is called and we cache some data about the call" when I feel they can that complexity when they get to the usage of that data.

but I could go either way